### PR TITLE
community[minor]: vercel kv graph checkpointer

### PR DESCRIPTION
**Pull request by chentschel**
Adds checkpointer for langgraph on vercel KV . cc/@jacoblee93

### CODE REVIEW

File: libs/langchain-azure-dynamic-sessions/package.json [hunk 1]
```diff
@@ -40,7 +40,7 @@
   "dependencies": {
     "@azure/identity": "^4.2.0",
     "@langchain/core": "~0.2",
-    "uuid": "^9.0.1"
+    "uuid": "^10.0.0"
```
Comment (jacoblee93) [inline libs/langchain-azure-dynamic-sessions/package.json:43]: This is probably a good idea, but can we do the individual integration packages in a separate PR?

File: libs/langchain-community/package.json [hunk 2]
```diff
@@ -239,6 +240,7 @@
     "@google-cloud/storage": "^6.10.1 || ^7.7.0",
     "@gradientai/nodejs-sdk": "^1.2.0",
     "@huggingface/inference": "^2.6.4",
+    "@langchain/langgraph": "^0.0.25",
```
Comment (jacoblee93) [inline libs/langchain-community/package.json:243]: Let's set this dep to `~0.0.25` so that people can automatically resolve to versions up to `0.1.0`

File: libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts
Comment (jacoblee93) [inline libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts:1]: Let's put under `langchain-community/src/langgraph/checkpointers/vercel_kv.ts`

Then import will be from under `@langchain/community/checkpointers/vercel_kv`.

Can you also add to `community/langchain.config.js`?

https://github.com/langchain-ai/langchainjs/blob/main/.github/contributing/INTEGRATIONS.md

**Comment by jacoblee93** [comment]
Thanks for this!

Looks good at a high-level, can we do the `uuid` version updates separately though?

I can patch all this feedback if you don't have time in the next few days.

File: libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
+import { VercelKV, createClient } from "@vercel/kv";
+
+import { RunnableConfig } from "@langchain/core/runnables";
+import {
+  BaseCheckpointSaver,
+  Checkpoint,
+  CheckpointMetadata,
+  CheckpointTuple,
+} from "@langchain/langgraph";
```
Comment (jacoblee93) [inline libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts:9]: This should import from `@langchain/langgraph/web`

File: libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+import { SerializerProtocol } from "../serde/base.js";
```
Comment (jacoblee93) [inline libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts:11]: Should import from `@langchain/langgraph/web` as well

**Comment by jacoblee93** [comment]
Blocked on https://github.com/langchain-ai/langgraphjs/pull/237/ and a version bump

File: libs/langchain-azure-dynamic-sessions/package.json (hunk 1 above)
Comment (chentschel) [reply libs/langchain-azure-dynamic-sessions/package.json]: Done. Added a bump of the package at https://github.com/langchain-ai/langchainjs/pull/5969.

File: libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts
Comment (chentschel) [reply libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts:1]: 👍

File: libs/langchain-community/package.json (hunk 2 above)
Comment (chentschel) [reply libs/langchain-community/package.json]: 👍

File: libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts (hunk 3 above)
Comment (chentschel) [reply libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts]: 👍

File: libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts (hunk 4 above)
Comment (chentschel) [reply libs/langchain-community/src/langgraph_checkpointers/vercel_kv.ts]: 👍

**Comment by chentschel** [comment]
> Blocked on [langchain-ai/langgraphjs#237](https://github.com/langchain-ai/langgraphjs/pull/237) and a version bump

Can be merged.

**Comment by jacoblee93** [comment]
Broadly LGTM, @nfcampos for a second look?

File: libs/langchain-community/src/langgraph/checkpointers/vercel_kv.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    const row: KVRow = {
+      checkpoint: this.serde.stringify(checkpoint),
+      metadata: this.serde.stringify(metadata),
+    };
+
+    // Save the checkpoint and the last checkpoint
+    await Promise.all([
+      this.kv.set(`${thread_id}:${checkpoint.id}`, row),
+      this.kv.set(`${thread_id}:last`, row),
```
Comment (jacoblee93) [inline libs/langchain-community/src/langgraph/checkpointers/vercel_kv.ts:142]: nit: Do this in an atomic operation?

File: libs/langchain-community/src/langgraph/checkpointers/vercel_kv.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  ): AsyncGenerator<CheckpointTuple> {
+    const thread_id = config.configurable?.thread_id;
+
+    // LUA script to get keys excluding those starting with "last"
+    const luaScript = `
+      local prefix = ARGV[1] .. ':'
+      local keys = redis.call('keys', prefix .. '*')
+      local result = {}
+      for _, key in ipairs(keys) do
+        if string.sub(key, string.len(prefix) + 1, string.len(prefix) + 4) ~= 'last' then
```
Comment (jacoblee93) [inline libs/langchain-community/src/langgraph/checkpointers/vercel_kv.ts:81]: nit: Use `:last` as a suffix

File: libs/langchain-community/src/langgraph/checkpointers/vercel_kv.ts (hunk 5 above)
Comment (chentschel) [reply libs/langchain-community/src/langgraph/checkpointers/vercel_kv.ts]: Great catch!. Done!.

File: libs/langchain-community/src/langgraph/checkpointers/vercel_kv.ts (hunk 6 above)
Comment (chentschel) [reply libs/langchain-community/src/langgraph/checkpointers/vercel_kv.ts]: I optimized the keys lookup and added the suggested check on `:last`

**Comment by jacoblee93** [comment]
Thank you! Docs can follow.

**Comment by jacoblee93** [comment]
Will ship a new version tonight! Thanks again!
